### ISSUE TITLE
native installation on Ubuntu 18.04 fails due to broken system pip

### ISSUE THREAD

**Issue by bertsky** [OCR-D/ocrd_all#266]
I don't know why we did not detect this earlier, but on a standard Ubuntu 18.04, after all system dependencies have been met (`deps-ubuntu`), a simple `make all` fails:

```
Traceback (most recent call last):
  File "/local/ocr-d/ocrd_all.new/venv/lib/python3.6/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/local/ocr-d/ocrd_all.new/venv/lib/python3.6/site-packages/pip/commands/install.py", line 290, in run
    with self._build_session(options) as session:
  File "/local/ocr-d/ocrd_all.new/venv/lib/python3.6/site-packages/pip/basecommand.py", line 69, in _build_session
    if options.cache_dir else None
  File "/usr/lib/python3.6/posixpath.py", line 80, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not int
Makefile:145: recipe for target '/local/ocr-d/ocrd_all.new/venv/bin/pip3' failed
```

This is of course pip version 9.0.1 preinstalled from the system. Since it is necessarily the first version that any venv will get set up with, and the _same error precludes even updating pip itself_, this is a pretty severe oversight both on Python's and on Ubuntu's side.

The only workaround I know is this:

    pip install --no-cache -U pip

I recommend adding this to our venv initialization recipe:

https://github.com/OCR-D/ocrd_all/blob/93c517b9c9ad79c78e6dfa37a7d0048f6393fa11/Makefile#L145

**Comment by stweil** [comment]
I cannot reproduce this issue. Making a virtual environment and upgrading pip works fine in my installation of Ubuntu bionic:

```
root@ub-blade-01:/# python3 -m venv venv
root@ub-blade-01:/# source venv/bin/activate
(venv) root@ub-blade-01:/# pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.1)
pkg-resources (0.0.0)
setuptools (32.3.1)
(venv) root@ub-blade-01:/# pip install --upgrade pip setuptools
Collecting pip
  Downloading https://files.pythonhosted.org/packages/47/ca/f0d790b6e18b3a6f3bd5e80c2ee4edbb5807286c21cdd0862ca933f751dd/pip-21.1.3-py3-none-any.whl (1.5MB)
    100% |################################| 1.6MB 687kB/s 
Collecting setuptools
  Downloading https://files.pythonhosted.org/packages/bd/25/5bdf7f1adeebd4e3fa76b2e2f045ae53ee208e40a4231ad0f0c3007e4353/setuptools-57.4.0-py3-none-any.whl (819kB)
    100% |################################| 819kB 1.2MB/s 
Installing collected packages: pip, setuptools
  Found existing installation: pip 9.0.1
    Uninstalling pip-9.0.1:
      Successfully uninstalled pip-9.0.1
  Found existing installation: setuptools 32.3.1
    Uninstalling setuptools-32.3.1:
      Successfully uninstalled setuptools-32.3.1
Successfully installed pip-21.1.3 setuptools-57.4.0
```

**Comment by stweil** [comment]
Wouldn't the Docker builds fail similarly if we had a problem?

**Comment by bertsky** [comment]
> Wouldn't the Docker builds fail similarly if we had a problem?

The Ubuntu Docker image `ubuntu:18.04` does not have that problem – these have probably been fixed by the maintainers already. It does happen with native 18.04.3 and 18.04.4 though:

    python3 -m venv venv
    . venv/bin/activate
    pip list
    pip install -U pip

```
pip (9.0.1)
pkg-resources (0.0.0)
setuptools (39.0.1)
Exception:
Traceback (most recent call last):
  File "venv/lib/python3.6/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "venv/lib/python3.6/site-packages/pip/commands/install.py", line 290, in run
    with self._build_session(options) as session:
  File "venv/lib/python3.6/site-packages/pip/basecommand.py", line 69, in _build_session
    if options.cache_dir else None
  File "/usr/lib/python3.6/posixpath.py", line 80, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not int
```

The same happens with Python 3.7:

    apt install python3.7-venv
    python3.7m -m venv venv
    . venv/bin/activate
    pip install -U pip

**Comment by stweil** [comment]
> It does happen with native 18.04.3 and 18.04.4 though

Then why don't I get it with any of my Ubuntu 18.04 installations?

There seems to be a difference: my installation shows setuptools (32.3.1), yours has setuptools (39.0.1) initially.

**Comment by bertsky** [comment]
> Why don't I get it with any of my Ubuntu 18.04 installations? They usually have all updates and fixes which are available for that release applied.

Perhaps you a patch version older than 3 or 4?

> There seems to be a difference: my installation shows setuptools (32.3.1), yours has setuptools (39.0.1) initially. Where does that come from?

Mine is from `bionic main`:

```
Package: python3-setuptools
Version: 39.0.1-2
APT-Sources: http://archive.ubuntu.com/ubuntu bionic/main amd64 Packages
```

Which makes me wonder where _your_ setuptools is from...

Interestingly though, all apt versions I have seem to be the exact same in a `ubuntu:18.04` Docker image (where again, the error does _not_ surface).

```sh
# dpkg-query -l python3-*| grep ^ii
ii  python3-asn1crypto            0.24.0-1                  all          Fast ASN.1 parser and serializer (Python 3)
ii  python3-cffi-backend          1.11.5-1                  amd64        Foreign Function Interface for Python 3 calling C code - runtime
ii  python3-crypto                2.6.1-8ubuntu2            amd64        cryptographic algorithms and protocols for Python 3
ii  python3-cryptography          2.1.4-1ubuntu1.4          amd64        Python library exposing cryptographic recipes and primitives (Python 3)
ii  python3-dbus                  1.2.6-1                   amd64        simple interprocess messaging system (Python 3 interface)
ii  python3-dev                   3.6.7-1~18.04             amd64        header files and a static library for Python (default)
ii  python3-distutils             3.6.9-1~18.04             all          distutils package for Python 3.x
ii  python3-gi                    3.26.1-2ubuntu1           amd64        Python 3 bindings for gobject-introspection libraries
ii  python3-idna                  2.6-1                     all          Python IDNA2008 (RFC 5891) handling (Python 3)
ii  python3-keyring               10.6.0-1                  all          store and access your passwords safely - Python 3 version of the package
ii  python3-keyrings.alt          3.0-1                     all          alternate backend implementations for python3-keyring
ii  python3-lib2to3               3.6.9-1~18.04             all          Interactive high-level object-oriented language (2to3, version 3.6)
ii  python3-minimal               3.6.7-1~18.04             amd64        minimal subset of the Python language (default python3 version)
ii  python3-pip                   9.0.1-2.3~ubuntu1.18.04.5 all          Python package installer
ii  python3-pkg-resources         39.0.1-2                  all          Package Discovery and Resource Access using pkg_resources
ii  python3-secretstorage         2.3.1-2                   all          Python module for storing secrets - Python 3.x version
ii  python3-setuptools            39.0.1-2                  all          Python3 Distutils Enhancements
ii  python3-six                   1.11.0-2                  all          Python 2 and 3 compatibility library (Python 3 interface)
ii  python3-venv                  3.6.7-1~18.04             amd64        pyvenv-3 binary for python3 (default python3 version)
ii  python3-wheel                 0.30.0-0.2                all          built-package format for Python
ii  python3-xdg                   0.25-4ubuntu1.1           all          Python 3 library to access freedesktop.org standards
# dpkg-query -l python-*| grep ^ii
ii  python-pip-whl           9.0.1-2.3~ubuntu1.18.04.5 all          Python package installer
```

I already see a difference in the generated `venv/bin/pip` though:

- my native systems uses `from pip import main`
- the Docker container uses `from pip._internal.cli.main import main`

I can even account for that difference, because my native system additionally has `python3-virtualenv` installed. If I fetch that in Docker, I do get the same `from pip import main` – but still no upgrade error in a fresh venv.

This has me completely puzzled. 

Still, since at least _some_ systems are affected, and adding `--no-cache` to the initial venv creation does not add any cost, I stand by #267. We cannot always afford to find the true cause.

**Comment by stweil** [comment]
> Which makes me wonder where _your_ setuptools is from...

I have not installed python3-setuptools.

**Comment by bertsky** [comment]
> I have not installed python3-setuptools.

Ok, but that would _still_ give you 39.0.1 once you are in the venv. 

(I have tried with and without installing system packages `python3-pip` and `python3-setuptools` – it does not matter.)

**Comment by stweil** [comment]
Ok, I had Ubuntu 18.04 (no patch level) initially. That works fine.

Now I added more apt sources for bionic-updates and bionic-security. That gives me 18.04.5 and the newer version of setuptools. And I still don't get the problem which you experience.

**Comment by stweil** [comment]
@bertsky, maybe your local pip cache is damaged. That can happen for any package. Instead of disabling the cache, it would be much better to clear your local cache.

**Comment by bertsky** [comment]
> maybe your local pip cache is damaged. That can happen for any package. Instead of disabling the cache, it would be much better to clear your local cache.

Indeed, that _was_ the culprit. To be precise: my cache was simply **deactivated**. And this happened (on both Ubuntu systems!) because I ran `pip install nvidia-pyindex` once (to get the Nvidia PyPI server for nvidia-tensorflow). I did not notice at the time, but that completely overwrote my ~/.pip/pip.conf ...

Of course, that's still a bug on pip's side, but I don't care at this point.

Thanks @stweil, I'll happily close then.
